pkg/controller/elasticsearch/driver/pvc.go

```diff
@@ -0,0 +1,122 @@
+		return err
+	}
+	return deleteUnusedPVCs(k8sClient, pvcs.Items, actualStatefulSets, expectedStatefulSets)
+}
+
+// reconcilePVCsOwnerRef ensures PVCs created for this Elasticsearch cluster have an owner ref set to
+// the Elasticsearch resource, so they are deleted automatically upon Elasticsearch deletion.
+// A subtle race condition exists: users may still end up with leftover PVCs if the Elasticsearch resource
+// gets deleted right after creation or update, before this is called.
+func reconcilePVCsOwnerRef(
+	k8sClient k8s.Client,
+	scheme *runtime.Scheme,
+	es v1alpha1.Elasticsearch,
+	pvcs []corev1.PersistentVolumeClaim,
+) error {
+	for _, pvc := range pvcs {
+		if err := setPVCOwnerRef(k8sClient, scheme, es, pvc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// setPVCOwnerRef sets an owner reference targeting es on the given pvc, if not already set.
+func setPVCOwnerRef(
+	k8sClient k8s.Client,
+	scheme *runtime.Scheme,
+	es v1alpha1.Elasticsearch,
+	pvc corev1.PersistentVolumeClaim,
+) error {
+	for _, ref := range pvc.OwnerReferences {
+		if ref.Name == es.Name {
+			// already set, nothing to do
+			return nil
+		}
+	}
+	log.V(1).Info("Setting PersistentVolumeClaim owner reference",
+		"namespace", es.Namespace,
+		"es_name", es.Name,
+		"pvc_name", pvc.Name,
+	)
+	if err := controllerutil.SetControllerReference(&es, &pvc, scheme); err != nil {
+		return err
+	}
+	return k8sClient.Update(&pvc)
+}
+
+// deleteUnusedPVCs deletes PVC resources that are not required anymore for this cluster.
+func deleteUnusedPVCs(
+	k8sClient k8s.Client,
+	pvcs []corev1.PersistentVolumeClaim,
+	actualStatefulSets sset.StatefulSetList,
+	expectedStatefulSets sset.StatefulSetList,
+) error {
+	for _, pvc := range pvcsToRemove(pvcs, actualStatefulSets, expectedStatefulSets) {
+		log.Info("Deleting PVC", "namespace", pvc.Namespace, "pvc_name", pvc.Name)
+		if err := k8sClient.Delete(&pvc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// pvcsToRemove filters the given pvcs to ones that can be safely removed based on Pods
+// of actual and expected StatefulSets.
+func pvcsToRemove(
+	pvcs []corev1.PersistentVolumeClaim,
+	actualStatefulSets sset.StatefulSetList,
+	expectedStatefulSets sset.StatefulSetList,
+) []corev1.PersistentVolumeClaim {
+	// Build the list of PVCs from both actual & expected StatefulSets (may contain duplicate entries).
+	// The list may contain PVCs for Pods that do not exist (eg. not created yet), but does not
+	// consider Pods in the process of being deleted (but not deleted yet), since already covered
+	// by checking expectations earlier in the process.
+	// Then, just return existing PVCs that are not part of that list.
+	expectedPVCs := append(actualStatefulSets.PVCNames(), expectedStatefulSets.PVCNames()...)
+	var toRemove []corev1.PersistentVolumeClaim
+	for _, pvc := range pvcs {
+		if !stringsutil.StringInSlice(pvc.Name, expectedPVCs) {
```

Comment:
Use map (set) instead?

Reply:
That's a small optimization given the relatively small (<1k items) size of our arrays here, but yeah, let's do it! 👍